### ISSUE TITLE
C# client .NET 3.5

### ISSUE THREAD

**Issue by jvdnbus** [krpc/krpc#392]
Do you believe it would be possible to build the C# client for .NET 3.5?

**Comment by djungelorm** [comment]
Should be possible. The only non-.NET 3.5 part of the is its use of `System.Tuple`, so that would have to be changed to some sort of custom tuple class. Google's Protocol Buffers library also requires NET 4.5, however I have already ported this library to .NET 3.5 (as that's what is required for it to run with KSP) so could release that along with the .NET 3.5 client.

**Comment by jvdnbus** [comment]
I've been looking into refactoring it for .NET 3.5 myself but I'm not that familiar with bazel. The reason is because I want to use the client in KSP, as well, to connect two KSP instances together.

**Comment by djungelorm** [comment]
That's a cool use case I hadn't considered!

**Comment by jvdnbus** [comment]
Do you know as to why this won't work?

https://github.com/jvdnbus/krpc/commit/72168d49216518bcec93a40da0ce37c0c17c633e#commitcomment-21870883

**Comment by djungelorm** [comment]
Your changes are a bit far reaching - you just want to build the C# client against the KSP .NET assemblies - these changes here should work: https://github.com/krpc/krpc/commit/2d851c0148e826e1b47bdc7b113487653f653865

When I build it I get the expected errors to do with System.Tuple not existing.

**Comment by jvdnbus** [comment]
Ah, yes. I substituted the System.Tuple with a copy of the KRPC.Utils.Tuple class now, and adapted the auto-generated service classes. However, when I try to build now, it tries to build the server ServiceDefinitions, which fails.

```
ERROR: ~/krpc/krpc/tools/ServiceDefinitions/BUILD:38:1: error executing shell command: 'mcs -target:library -debug -optimize+ -noconfig -nostdlib -out:bazel-out/host/bin/tools/ServiceDefinitions/KRPC.dll -doc:bazel-out/host/bin/tools/ServiceDefinitions/KRPC.xml -warn:4 -warnaserror+ s...' failed: Process terminated by signal 6 [sandboxed].
* Assertion at domain.c:718, condition `mono_defaults.monotype_class != 0' not met
...
```

**Comment by jvdnbus** [comment]
Nevermind, I gave up on building it on Linux because it was a pain to get .NET 3.5. I just went with an archived Windows Mono installation, version 3.12. Now I'm simply building it through Xamarin and every error I encountered, I fixed. The issues I encountered were the following:
- As mentioned earlier, System.Tuple is missing, so I replaced it with a copy of your implementation under KRPC.Utils. Refactored it where necessary. This meant adapting the bazel generated files (dirty but I didn't feel like building those and potentially run into more issues with older versions).
- System.Collections.Generic.ISet<T> does not exist in .NET 3.5, I used HashSet<T> instead, which is under the same namespace, and refactored the Encoder as well.

It seems to work in-game at first glance. Will keep you updated.

**Comment by djungelorm** [comment]
Sounds like you are missing a bunch of auto-generated files. These can either be built using bazel (following the instructions here http://krpc.github.io/krpc/compiling.html) or from a genfiles archive (which is attached to each release on github). The latest is here: https://github.com/krpc/krpc/releases/download/v0.3.8/krpc-genfiles-0.3.8.zip

**Comment by EoPrime** [comment]
Thank you!
bazel didnt like me very much but helped somehow, now with the archive and changes mentioned by jvdnbus it did finally compile in VS.

**Comment by djungelorm** [comment]
Cool. I don't think bazel works that well on Windows...

FYI I plan to release 3.5 compatible binaries in (hopefully) the next and future releases.

**Comment by djungelorm** [comment]
I've added support for .NET 3.5 - will be released in v0.3.10.

A pre-release is available at the following link if you can't wait for that ;)
https://krpc.s3.amazonaws.com/deploy/feature/csharp-client-mono-3.5/925.1/client/krpc-csharp-0.3.10-11-g41c0d6a.zip

**Comment by EoPrime** [comment]
Thank you very much!

My version had problems in the encoder.
